### PR TITLE
*: add e2e elect, lock test

### PR DESCRIPTION
**Pull request by gyuho**

### CODE REVIEW

**Comment by gyuho** [comment]
1. `../bin/etcdctl --endpoints localhost:20000,localhost:20002,localhost:20004 --dial-timeout 7s elect a b` outputs `A is holding election "a/1460161307137087759"`
3. `../bin/etcdctl --endpoints localhost:20000,localhost:20002,localhost:20004 --dial-timeout 7s elect a --listen` outputs `B is listening`
4. `A is killed`
5. `../bin/etcdctl --endpoints localhost:20000,localhost:20002,localhost:20004 --dial-timeout 7s elect a b2` outputs `ctl_v3_election_test.go:78: timed out 3s`

@heyitsanthony Could you quickly take a look at my test? It works well by step 3, but trying to killing the client holding the election doesn't seem to work. I tried to send `os.Kill` or `syscall.SIGTERM` and tried to get notified in `elect` command but the command doesn't seem to catch that signal. I will investigate more later today.

Manually running `elect` command and interrupt with `ctrl + c` works as expected. Thanks!

**Comment by heyitsanthony** [comment]
@gyuho `elect` will only catch SIGINT because golang only guarantees support for `Interrupt` and `Kill`-- try SIGINT instead of SIGTERM?

**Comment by gyuho** [comment]
@heyitsanthony ok will try! thanks!

**Comment by gyuho** [comment]
Ended up adding `os.Getpid()` and `os.Getppid()` with debug mode:

<br><hr>
###### Example 1

```go
// sig.go
package main

import (
	"fmt"
	"os"
	"os/signal"
	"time"
)

func main() {
	fmt.Printf("waiting with pid %d / ppid %d\n", os.Getpid(), os.Getppid())

	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt, os.Kill)
	go func() {
		s := <-c
		fmt.Println("got:", s)
	}()

	for {
		time.Sleep(2 * time.Second)
	}
}
```

`go build sig.go` and `go run main.go`:

```go
// main.go
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/coreos/etcd/pkg/expect"
)

func main() {
	ex, err := expect.NewExpect("./sig")
	if err != nil {
		log.Fatal(err)
	}
	l, err := ex.ExpectFunc(func(txt string) bool { return true })
	fmt.Println(l, err)
	// waiting with pid 3614 / ppid 3609

	ep, err := ex.Pid()
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("expect pid:", ep, err) // expect pid: 3614
	fmt.Println("signalling...:", ex.Signal(os.Interrupt))

	l, err = ex.ExpectFunc(func(txt string) bool { return true })
	fmt.Println(l, err)
	// got: interrupt
}
```

This works as expected. `expect` package returns correct PID of the sub-process and `Signal` method worked to send `os.Interrupt` to the sub-process.


<br><hr>
###### Example 2

Run this PR with more logging.

```
pid from command: pid: 9039 / ppid: 9038
pid from expect: 9038
```

The pid from `expect` pkg returns the ppid of the actual sub-process in this case.
So that's why I added debug line to print the actual pid using `os.Getpid()`.

Any idea? I will investigate more.

**Comment by heyitsanthony** [comment]
@gyuho the only thing that sticks out is it should be calling `expect.Close()` after signalling instead of `expect.Stop()`.
Likewise, https://github.com/coreos/etcd/pull/5014/files#diff-fa8f23c297cf9f5e78665608450ff24aR68 looks like it races since it doesn't Wait on the first process's termination.

**Comment by gyuho** [comment]
@heyitsanthony Will try that. Thanks!

**Comment by gyuho** [comment]
@heyitsanthony PTAL. Now it `Close` after signalling to wait for its termination. Add more test steps to ensure duplicate locks get blocked.

File: etcdctl/ctlv3/command/elect_command.go [hunk 1]
```diff
@@ -37,6 +39,7 @@ func NewElectCommand() *cobra.Command {
 		Run:   electCommandFunc,
 	}
 	cmd.Flags().BoolVarP(&electListen, "listen", "l", false, "observation mode")
+	cmd.Flags().BoolVarP(&electDebug, "debug", "d", false, "debug mode")
```
Comment (heyitsanthony) [inline etcdctl/ctlv3/command/elect_command.go]: nuke this flag

File: etcdctl/ctlv3/command/lock_command.go [hunk 2]
```diff
@@ -25,13 +26,16 @@ import (
...
+var lockDebug bool
+
 // NewLockCommand returns the cobra command for "lock".
 func NewLockCommand() *cobra.Command {
 	c := &cobra.Command{
 		Use:   "lock <lockname>",
 		Short: "lock acquires a named lock",
 		Run:   lockCommandFunc,
 	}
+	c.Flags().BoolVarP(&lockDebug, "debug", "d", false, "debug mode")
```
Comment (heyitsanthony) [inline etcdctl/ctlv3/command/lock_command.go]: nuke this flag

File: pkg/expect/expect.go [hunk 3]
```diff
@@ -143,6 +143,22 @@ func (ep *ExpectProcess) close(kill bool) error {
...
+// Signal sends os.Signal to the expect process.
+func (ep *ExpectProcess) Signal(sig os.Signal) error {
+	if ep.cmd == nil {
+		return ep.err
+	}
+	return ep.cmd.Process.Signal(sig)
+}
+
+// Pid returns the pid of the expect process.
+func (ep *ExpectProcess) Pid() (int, error) {
```
Comment (heyitsanthony) [inline pkg/expect/expect.go]: is this necessary if `Signal` is already available?

File: e2e/ctl_v3_lock_test.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+		cx.t.Fatal(err)
+	}
+	if es1 == es2 {
+		cx.t.Fatalf("expected different election name, got %q == %q", es1, es2)
+	}
+}
+
+// ctlV3Lock spawns elect command without stopping(or closing) it.
+// Make sure to stop it to prevent leaky goroutine.
+func ctlV3Lock(cx ctlCtx, name string, timeout time.Duration) (*expect.ExpectProcess, int, string, error) {
```
Comment (heyitsanthony) [inline e2e/ctl_v3_lock_test.go]: probably not worth parameterizing the timeout; just set to 5s?

File: e2e/ctl_v3_lock_test.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	//
+	if err = syscall.Kill(bp, syscall.SIGINT); err != nil {
+		cx.t.Fatal(err)
+	}
+	if err = blocked.Close(); err != nil {
+		cx.t.Fatal(err)
+	}
+
+	// kill the node1 holding the lock
+	if err = syscall.Kill(pid1, syscall.SIGINT); err != nil {
```
Comment (heyitsanthony) [inline e2e/ctl_v3_lock_test.go]: why can't this use `expect.Signal`?

File: e2e/ctl_v3_lock_test.go (hunk 4 above)
Comment (heyitsanthony) [inline e2e/ctl_v3_lock_test.go]: also is the pid necessary?

File: e2e/ctl_v3_elect_test.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+	}
+	if err = syscall.Kill(bp, syscall.SIGINT); err != nil {
+		cx.t.Fatal(err)
+	}
+	if err = blocked.Close(); err != nil {
+		cx.t.Fatal(err)
+	}
+
+	// kill the node holding the election(leadership)
+	if err = syscall.Kill(pid1, syscall.SIGINT); err != nil {
```
Comment (heyitsanthony) [inline e2e/ctl_v3_elect_test.go]: `node1.Signal()`?

File: e2e/ctl_v3_elect_test.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+	if es1 == es3 {
+		cx.t.Fatalf("expected different election name, got %q == %q", es1, es3)
+	}
+
+	// ensure new election, proposal get watched
+	es4, err := listener.ExpectFunc(tFunc)
+	if err != nil {
+		cx.t.Fatal(err)
+	}
+	es4 = strings.Replace(es4, "\r\n", "", -1)
```
Comment (heyitsanthony) [inline e2e/ctl_v3_elect_test.go]: what happened to the newlines in `es3`?

File: etcdctl/ctlv3/command/elect_command.go (hunk 1 above)
Comment (gyuho) [inline etcdctl/ctlv3/command/elect_command.go]: I will rework on this tonight. The reason why I put this flag is to get the pid from the process printed as explained in https://github.com/coreos/etcd/pull/5014#issuecomment-208437155. Not sure why `Process.Pid` returns the `ppid` of the process. Thanks!

**Comment by gyuho** [comment]
@heyitsanthony PTAL. Removed `debug` mode. And found a way to `SIGINT` the commands. Problem is:

- `expect.Pid` method returns the ppid, not the actual pid of the actual command process
- so didn't work by sending signal to the ppid
- now it finds the pgid and kills its children

I couldn't figure out why it still returns ppid though.

**Comment by heyitsanthony** [comment]
@gyuho I can't reproduce the ppid thing:
```go
package main

import (
        "fmt"
        "os/exec"
        "os"
)

func main() {
        cmd := exec.Command("/bin/cat", "/proc/self/stat")
        if cmd == nil {
                fmt.Println("failed to run cmd")
                os.Exit(1)
        }
        cmd.Stdout = os.Stdout
        if err := cmd.Run(); err != nil {
                fmt.Println(err)
                os.Exit(2)
        }

        fmt.Println("cmd proc:", cmd.Process.Pid)
        fmt.Println("this proc:", os.Getpid(), os.Getppid())
}
```
as output I get:
```
5414 (cat) R 5409 5409 15514 34856 5409 4210688 53 0 0 0 0 0 0 0 20 0 1 0 164330411 4423680 184 18446744073709551615 4194304 4239972 140735787040768 140735787040120 139904808073712 0 0 0 0 0 0 0 17 3 0 0 0 0 0 6340112 6341396 33009664 140735787048288 140735787048313 140735787048313 140735787053039 0
cmd proc: 5414
this proc: 5409 15514
```
The 5414 from `cmd.Process.Pid` matches pid field of `/proc/self/stat`. The `/proc/self/stat` ppid field matches the pid of the calling process 5409.

I added
```
// Signal sends a signal to the expect process                                                        
func (ep *ExpectProcess) Signal(sig os.Signal) error { return ep.cmd.Process.Signal(sig) }
```

and replaced the signal stuff in the lock test with usages like:
`if err = blocked.Signal(syscall.SIGINT); err != nil {`

and it seems to work...

**Comment by gyuho** [comment]
@heyitsanthony Thanks, but the pids do not match if we do the same thing to the lock command for some reason. I pushed another commit to use your `Signal` method and doesn't seem to work :(

**Comment by heyitsanthony** [comment]
@gyuho I tested the expect.Signal version on my linux machine and it works fine. Is it breaking locally for you as well or is it only semaphore? I'll investigate further...

**Comment by gyuho** [comment]
@heyitsanthony That's strange... I am on linux too.

```
Linux gyuho 4.4.0-18-generic #34-Ubuntu SMP Wed Apr 6 14:01:02 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

with go 1.6. It might be just me and CI.

**Comment by gyuho** [comment]
Closing in favor of https://github.com/coreos/etcd/pull/5056.

**Comment by heyitsanthony** [comment]
@gyuho #5056 is missing the election test though... do you want to do that one or should I pick it up?

**Comment by gyuho** [comment]
@heyitsanthony Yeah I am sending a new PR to test elect command. Thanks!
